Improvements to the configuration helpers

pingvinen opened the pull request:
Lately I have spotted some possible improvements to the configuration helper(s).

1. have had multiple cases where I would have liked to have been able to have a shared config value as a fallback - i.e. being able to do something like `Get("specific_key", "fallback_key", "default value")`
2. have implemented handling of comma separated values multiple times (get value as string, split it, clean it up then convert it to a list)
3. started thinking that maybe the `ConfigBase` is not really needed as all it contains is a method that uses the `IConfiguration` that it is given - but that logic could just as easily be implemented as an extension method on the `IConfiguration` as it does not need to be mockable (we would always mock the interface/class itself anyway)

This MR addresses all of the above + adds documentation of it :)

pingvinen committed:
Move ConfigBase.Get<T>(..) implementation to an ext.method

Also, [obsolete] the `ConfigBase` class, as it really is not needed
anymore.

pingvinen committed:
Add IConfiguration.Get<T> methods with multiple keys

pingvinen committed:
Add IConfiguration.GetCommaSeparatedList<T>(..)

pingvinen committed:
Add markdown linting rules

pingvinen committed:
Add documentation of the Configuration stuff

pingvinen commented:
Hmm.. the travis build seems to be broken. I will have to look into that.

IceAndSnow commented:
I think it looks good. How about also adding a multi-keyed `GetCommaSeparatedList`?

```
        public static List<T> GetCommaSeparatedList<T>(this IConfiguration config, IEnumerable<string> keys, List<T> defaultValue)
        {
            var value = config.Get<string>(keys, null);

            if (string.IsNullOrEmpty(value))
            {
                return defaultValue;
            }

            return value
                .Split(new[] {","}, StringSplitOptions.RemoveEmptyEntries)
                .Where(x => !string.IsNullOrWhiteSpace(x))
                .Select(x => ConvertOrDefault(x.Trim(), default(T)))
                .ToList();
        }
```

pingvinen committed:
Fail travis build on first error

pingvinen commented:
> Hmm.. the travis build seems to be broken. I will have to look into that.

Apparently it has nothing to do with Travis :) It is the `mkdocs` version versus `python` version battle - it is super brittle apparently.

pingvinen committed:
Add multi-key GetCommaSeparatedList<T>

pingvinen commented:
> I think it looks good. How about also adding a multi-keyed `GetCommaSeparatedList`?

I have added overloads similar to the ones for `Get` - i.e. with 2 keys, 3 keys and X keys.

IceAndSnow commented on docs/RapidCore.Configuration/Configuration.md:
````diff
@@ -0,0 +1,99 @@
+}
+```
+
+In the config classes it is now possible to do `public string Host => Get("db_a:host", "db_shared:host", "no host")`.
````
**question (non-blocking):** This exact example only works when using `ConfigBase`, right?

I think it would be better, to instead provide an example not relying on the now obsolete `ConfigBase`, e.g:

```suggestion
In the config classes it is now possible to do `public string Host => configuration.Get("db_a:host", "db_shared:host", "no host")`.
```

nover replied:
@pingvinen should we get this question fixed?

pingvinen replied:
Yes

IceAndSnow commented on src/core/main/Configuration/ConfigurationGetExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// 2. empty values are removed
+        /// 3. each value is trimmed and converted (just like Get would do)
+        /// </summary>
+        public static List<T> GetCommaSeparatedList<T>(this IConfiguration config, IEnumerable<string> keys, List<T> defaultValue)
```
**praise:** I think this is a great addition :)

nover commented on src/core/main/Configuration/ConfigurationGetExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        /// <summary>
+        /// This is a convenience overload for GetCommaSeparatedList(string[] keys, T defaultValue)
+        /// </summary>
+        public static List<T> GetCommaSeparatedList<T>(this IConfiguration config, string key, List<T> defaultValue)
```
**suggestion**: rename to `GetFromCommaSeparatedList<T>(args)`

First time I read it I tought: "get these values and return as a comma separated list"

pingvinen replied:
Makes sense.. I will get that fixed.

pingvinen committed:
Rename GetCommaSeparatedList to GetFromCommaSeparatedList

pingvinen committed:
Restructure the configuration documentation a bit

Changed the order a bit to put more focus on the IConfiguration
extension method pattern we have introduced.

pingvinen commented:
I have incorporated your suggestions (@IceAndSnow and @nover) in code and documentation.

I have also restructured the documentation of the config stuff a bit to put focus on the new pattern.

IceAndSnow commented on src/core/main/Configuration/ConfigurationGetExtensions.cs:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        /// <summary>
+        /// This is a convenience overload for GetFromCommaSeparatedList(string[] keys, T defaultValue)
```
**nitpick**: The default value is type `List<T>`

This is also the case for the next two extension methods.

pingvinen replied:
That should be fixed now.

IceAndSnow reviewed (approved):
This looks good as far as I can see
